Unit and end-to-end tests are flakey

They fail from time to time. e.g. https://circleci.com/gh/fluxcd/flux/8957

@hiddeco @squaremo @stefanprodan please add more cases here (if you see them) so that I can fix them.

I managed to make it fail https://circleci.com/gh/fluxcd/flux/9008

```
Running: common case
calling updateWorkloadContainer(manifest, extra:deployment/pr-assigner, pr-assigner, quay.io/weaveworks/pr-assigner:master-1234567)
Took: 1.1 seconds
Running: new version like number
calling updateWorkloadContainer(manifest, default:deployment/fluxy, fluxy, weaveworks/fluxy:1234567)
Took: 1.1 seconds
Running: old version like number
calling updateWorkloadContainer(manifest, default:deployment/fluxy, fluxy, weaveworks/fluxy:master-a000001)
Took: 1.0 seconds
Running: name label out of order
calling updateWorkloadContainer(manifest, monitoring:deployment/grafana, grafana, quay.io/weaveworks/grafana:master-37aaf67)
Took: 1.0 seconds
Running: version (tag) with dots
calling updateWorkloadContainer(manifest, sock-shop:deployment/front-end, front-end, weaveworksdemos/front-end:7f511af2d21fd601b86b3bed7baa6adfa9c8c669)
Took: 1.0 seconds
Running: minimal dockerhub image name
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx, nginx:1.10-alpine)
Took: 1.1 seconds
Running: reordered keys
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx, nginx:1.10-alpine)
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx2, nginx:1.10-alpine)
Took: 2.3 seconds
Running: from prod
calling updateWorkloadContainer(manifest, default:deployment/authfe, logging, quay.io/weaveworks/logging:master-123456)
Took: 1.1 seconds
Running: single quotes
calling updateWorkloadContainer(manifest, default:deployment/weave, weave, weaveworks/weave-kube:2.2.1)
Took: 1.0 seconds
Running: in multidoc
calling updateWorkloadContainer(manifest, hello:deployment/helloworld, helloworld, quay.io/weaveworks/helloworld:master-a000001)
Took: 1.0 seconds
Running: in kubernetes List resource
calling updateWorkloadContainer(manifest, hello:deployment/helloworld, helloworld, quay.io/weaveworks/helloworld:master-a000001)
Took: 1.2 seconds
Running: FluxHelmRelease (v1alpha2; simple image encoding)
calling updateWorkloadContainer(manifest, maria:fluxhelmrelease/mariadb, chart-image, bitnami/mariadb:10.1.33)
Took: 1.0 seconds
Running: FluxHelmRelease (v1alpha2; multi image encoding)
calling updateWorkloadContainer(manifest, maria:fluxhelmrelease/mariadb, mariadb, bitnami/mariadb:10.1.33)
Took: 1.0 seconds
Running: HelmRelease (v1beta1; image with port number)
calling updateWorkloadContainer(manifest, maria:helmrelease/mariadb, mariadb, localhost:5000/mariadb:10.1.33)
Took: 1.0 seconds
Running: HelmRelease (v1; with image map)
Took: 1.0 seconds
Running: initContainer
calling updateWorkloadContainer(manifest, default:deployment/weave, weave, weaveworks/weave-kube:2.2.1)
panic: test timed out after 1m0s
```

I am pretty sure docker gets stuck when executing `kubeyaml`. Why, I still don't know.

Here is another instance https://circleci.com/gh/fluxcd/flux/9013 . In this case it fails in another test-case:

```
Running: common case
calling updateWorkloadContainer(manifest, extra:deployment/pr-assigner, pr-assigner, quay.io/weaveworks/pr-assigner:master-1234567)
Took: 1.2 seconds
Running: new version like number
calling updateWorkloadContainer(manifest, default:deployment/fluxy, fluxy, weaveworks/fluxy:1234567)
Took: 1.1 seconds
Running: old version like number
calling updateWorkloadContainer(manifest, default:deployment/fluxy, fluxy, weaveworks/fluxy:master-a000001)
Took: 1.1 seconds
Running: name label out of order
calling updateWorkloadContainer(manifest, monitoring:deployment/grafana, grafana, quay.io/weaveworks/grafana:master-37aaf67)
Took: 1.0 seconds
Running: version (tag) with dots
calling updateWorkloadContainer(manifest, sock-shop:deployment/front-end, front-end, weaveworksdemos/front-end:7f511af2d21fd601b86b3bed7baa6adfa9c8c669)
Took: 1.1 seconds
Running: minimal dockerhub image name
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx, nginx:1.10-alpine)
Took: 1.2 seconds
Running: reordered keys
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx, nginx:1.10-alpine)
calling updateWorkloadContainer(manifest, default:deployment/nginx, nginx2, nginx:1.10-alpine)
Took: 2.0 seconds
Running: from prod
calling updateWorkloadContainer(manifest, default:deployment/authfe, logging, quay.io/weaveworks/logging:master-123456)
Took: 1.0 seconds
Running: single quotes
calling updateWorkloadContainer(manifest, default:deployment/weave, weave, weaveworks/weave-kube:2.2.1)
Took: 1.0 seconds
Running: in multidoc
calling updateWorkloadContainer(manifest, hello:deployment/helloworld, helloworld, quay.io/weaveworks/helloworld:master-a000001)
Took: 0.9 seconds
Running: in kubernetes List resource
calling updateWorkloadContainer(manifest, hello:deployment/helloworld, helloworld, quay.io/weaveworks/helloworld:master-a000001)
Took: 1.1 seconds
Running: FluxHelmRelease (v1alpha2; simple image encoding)
calling updateWorkloadContainer(manifest, maria:fluxhelmrelease/mariadb, chart-image, bitnami/mariadb:10.1.33)
Took: 1.0 seconds
Running: FluxHelmRelease (v1alpha2; multi image encoding)
calling updateWorkloadContainer(manifest, maria:fluxhelmrelease/mariadb, mariadb, bitnami/mariadb:10.1.33)
Took: 1.0 seconds
Running: HelmRelease (v1beta1; image with port number)
calling updateWorkloadContainer(manifest, maria:helmrelease/mariadb, mariadb, localhost:5000/mariadb:10.1.33)
Took: 1.0 seconds
Running: HelmRelease (v1; with image map)
panic: test timed out after 1m0s
```

OK. I think I know what's happening (and I feel a bit stupid about it).

TL;DR: we were being too conservative in the `-timeout` `go test` flag. And the flag doesn't work like I expected.

We set a timeout of 60s when running the tests. I had assumed that the timeout applied **per test** (i.e. per `t.Run()` invokation or `Test*` function). The documentation seems to indicate so:

```
    -timeout d
        If a test binary runs longer than duration d, panic.
        If d is 0, the timeout is disabled.
        The default is 10 minutes (10m).
```

It's not completely clear to me what _a binary_ means, but, in the Makefile pass all the flux go packages to `go test`:

```
test: test/bin/helm test/bin/kubectl test/bin/kustomize $(GENERATED_TEMPLATES_FILE)
	PATH="${PWD}/bin:${PWD}/test/bin:${PATH}" go test ${TEST_FLAGS} $(shell go list ./... | grep -v "^github.com/fluxcd/flux/vendor" | sort -u)
```


For instance, this fails
```
$ cd flux
$ PATH=$PATH:$PWD/bin/ time go test  -race -tags integration -timeout 60s github.com/fluxcd/flux/pkg/cluster/kubernetes/
[...]
=== RUN   TestWorkloadContainerUpdates/HelmRelease_(v1;_with_image_map)
panic: test timed out after 1m0s
[...]
FAIL	github.com/fluxcd/flux/pkg/cluster/kubernetes	60.078s
FAIL
```

But if we filter `TestWorkloadContainerUpdates` specifically it passes.

For instance, this fails
```
$ cd flux
$ PATH=$PATH:$PWD/bin/ time go test  -run TestWorkloadContainerUpdates -race -tags integration -timeout 60s github.com/fluxcd/flux/pkg/cluster/kubernetes/
[...]
--- PASS: TestWorkloadContainerUpdates (25.41s)
[...]
ok  	github.com/fluxcd/flux/pkg/cluster/kubernetes	(cached)
```

So, I think this started happening when adding new tests to `pkg/cluster/kubernetes/` in 18dd15e57db08c90260ece766caa7479a47b6a7d .

Also, this didn't surface all that much because of caching (the test was only re-run when the package was modified and if you got lucky, it run fast enough, getting cached again).

Since it's hard to control  `-timeout` granularity I will be generous with it.

Jeez, this took way longer than it should had.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -110,7 +110,8 @@ jobs:
           keys:
             - go-mod-{{ checksum "go.mod" }}
             - go-mod-
-      - run: make test TEST_FLAGS="-race -tags integration -timeout 60s"
+      # Note that the test timeout is per Go package
+      - run: make test TEST_FLAGS="-race -tags integration -timeout 5m"
       - run: make check-generated
       - run: make all
       - run: make e2e

--- a/Makefile
+++ b/Makefile
@@ -59,7 +59,7 @@ realclean: clean
 	rm -rf ./cache
 
 test: test/bin/helm test/bin/kubectl test/bin/kustomize $(GENERATED_TEMPLATES_FILE)
-	PATH="${PWD}/bin:${PWD}/test/bin:${PATH}" go test ${TEST_FLAGS} $(shell go list ./... | grep -v "^github.com/fluxcd/flux/vendor" | sort -u)
+	PATH="${PWD}/bin:${PWD}/test/bin:${PATH}" go test ${TEST_FLAGS} $(shell go list ./... | sort -u)
 
 e2e: lint-e2e test/bin/helm test/bin/kubectl test/e2e/bats $(GOBIN)/fluxctl build/.flux.done
 	PATH="${PWD}/test/bin:${PATH}" CURRENT_OS_ARCH=$(CURRENT_OS_ARCH) test/e2e/run.bash

